controller/api/destination/server.go

```diff
@@ -383,11 +425,38 @@ func getSvcID(k8sAPI *k8s.API, clusterIP string, log *logging.Entry) (*watcher.S
 	return service, nil
 }
 
-// getPod returns a pod that maps to the given IP address. The pod can either
+// getPodByHostname returns a pod that maps to the given hostname (or an
+// instanceID). The hostname is generally the prefix of the pod's DNS name;
+// since it may be arbitrary we need to look at the corresponding service's
+// Endpoints object to see whether the hostname matches a pod.
+func getPodByHostname(k8sAPI *k8s.API, hostname string, svcID watcher.ServiceID) (*corev1.Pod, error) {
+	ep, err := k8sAPI.Endpoint().Lister().Endpoints(svcID.Namespace).Get(svcID.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, subset := range ep.Subsets {
+		for _, addr := range subset.Addresses {
+			if addr.TargetRef == nil || addr.TargetRef.Kind != "Pod" {
```

Comment:
One thing to note is that this won't work for headless services with non-pod endpoints.  I can't think of a use-case for this off the top of my head, but I think we could probably support it without too much more effort.  However, it's probably best to leave that as a TODO or a github issue for now.

Reply:
Sounds good. I'll add a TODO and create an issue for it. 

I can think of two use-cases for this: if we have synthetic endpoints (such as when we want to build a connection directly to a gateway IP rather than go through `Get`) and with external IPs.